Grid/qcd/utils/BaryonUtils.h

```diff
@@ -249,4 +310,246 @@ void BaryonUtils<FImpl>::ContractBaryons_Sliced(const mobj &D1,
+
+  auto DuG = Du_spec * GammaB_nucl;
+  auto adjDu = g5 * adj(Du_xf) * g5;
+  auto adjDuG = adjDu * GammaB_nucl;
+  // Gamma^B * Ds * \gamma_\mu^L
+  auto GDsG = GammaB_sigma * Ds_xi * Gamma_H;
+  // Du * \gamma_\mu^L * (\gamma_5 * Dd^\dagger * \gamma_5)
+  auto DuGDd = Du_xi * Gamma_H * g5 * adj(Dd_xf) * g5;
+
+  for (int ie_n=0; ie_n < 6 ; ie_n++){
+    int a_n = epsilon[ie_n][0]; //a
+    int b_n = epsilon[ie_n][1]; //b
+    int c_n = epsilon[ie_n][2]; //c
+    for (int ie_s=0; ie_s < 6 ; ie_s++){
+      int a_s = epsilon[ie_s][0]; //a'
+      int b_s = epsilon[ie_s][1]; //b'
+      int c_s = epsilon[ie_s][2]; //c'
+      for (int i=0; i<Nc; i++){
+      for (int alpha_s=0; alpha_s<Ns; alpha_s++){
+      for (int beta_n=0; beta_n<Ns; beta_n++){
+      for (int tau=0; tau<Ns; tau++){
+      for (int gamma_s=0; gamma_s<Ns; gamma_s++){
+      for (int gamma_n=0; gamma_n<Ns; gamma_n++){
+        result()(gamma_s,gamma_n)() -= static_cast<Complex>(epsilon_sgn[ie_n] * epsilon_sgn[ie_s]) * GDsG()(alpha_s,tau)(b_s,i) * DuGDd()(alpha_s,beta_n)(a_s,b_n) * DuG()(gamma_s, beta_n)(c_s,a_n) * adjDu()(tau,gamma_n)(i,c_n);
+        result()(gamma_s,gamma_n)() -= static_cast<Complex>(epsilon_sgn[ie_n] * epsilon_sgn[ie_s]) * GDsG()(alpha_s,tau)(b_s,i) * DuGDd()(gamma_s,beta_n)(c_s,b_n) * adjDuG()(tau, beta_n)(i,a_n) * Du_spec()(alpha_s,gamma_n)(a_s,c_n);
+        result()(gamma_s,gamma_n)() += static_cast<Complex>(epsilon_sgn[ie_n] * epsilon_sgn[ie_s]) * GDsG()(alpha_s,tau)(b_s,i) * DuGDd()(alpha_s,beta_n)(a_s,b_n) * adjDuG()(tau, beta_n)(i,a_n) * Du_spec()(gamma_s,gamma_n)(c_s,c_n);
+        result()(gamma_s,gamma_n)() += static_cast<Complex>(epsilon_sgn[ie_n] * epsilon_sgn[ie_s]) * GDsG()(alpha_s,tau)(b_s,i) * DuGDd()(gamma_s,beta_n)(c_s,b_n) * DuG()(alpha_s, beta_n)(a_s,a_n) * adjDu()(tau,gamma_n)(i,c_n);
+      }}}}}}
+    }
+  }
+}
+
+
+template<class FImpl>
+template <class mobj>
+void BaryonUtils<FImpl>::Sigma_to_Nucleon(const PropagatorField &qq_xi,
+						 const PropagatorField &qq_xf,
+						 const mobj &Du_spec,
+						 const PropagatorField &qd_xf,
+						 const PropagatorField &qs_xi,
+				                 const Gamma Gamma_H,
+				                 const Gamma GammaB_sigma,
+		                 		 const Gamma GammaB_nucl,
+						 const int parity,
+						 const std::string op,
+						 SpinMatrixField &stn_corr)
+{
+  assert(parity==1 || parity == -1 && "Parity must be +1 or -1");
+
+  GridBase *grid = qs_xi.Grid();
+
+  auto vcorr= stn_corr.View();
+  auto vq_xi = qq_xi.View();
+  auto vq_xf = qq_xf.View();
+  auto vd_xf = qd_xf.View();
+  auto vs_xi = qs_xi.View();
+
+ // accelerator_for(ss, grid->oSites(), grid->Nsimd(), {
+  thread_for(ss,grid->oSites(),{
+    auto Dq_xi = vq_xi[ss];
+    auto Dq_xf = vq_xf[ss];
+    auto Dd_xf = vd_xf[ss];
+    auto Ds_xi = vs_xi[ss];
+    sobj result=Zero();
+    if(op == "Q1_NonEye"){
+      Sigma_to_Nucleon_Q1_NonEye_site(Dq_xi,Dq_xf,Du_spec,Dd_xf,Ds_xi,Gamma_H,GammaB_sigma,GammaB_nucl,parity,result);

```

Comment:
A design decision which I am not so sure about: The Eye function has only 4 Propagators as input, the NonEye ones 5. As I have only one function for both, the second argument for the Eye functions is ignored, and the name is misleading. (should be qq_loop).

I could just split these into two functions, SigmaToNucleonEye and SigmaToNucleonNonEye, that would be easy and less ambiguous (but longer).